### PR TITLE
Add the ability to 'presell' one-day badges at-door

### PR DESCRIPTION
**Pull request by kitsuta**
For this year, we wanted to allow anyone to buy any day's badge through our 'at-door' page.

This was trickier than expected, because our badge configuration is fairly limited - you can't have multiple 'attendee'-type badges with their own sets of prices. While this is a good argument for refactoring badge types into object-based entities, that was too much to tackle right now.

I'll explain the changes in-line.

### CODE REVIEW

File: uber/config.py [hunk 1]
```diff
@@ -165,8 +174,33 @@ def SUPPORTERS_ENABLED(self):
...
+        This populates the dropdown form on the /registration/register page with a list of badges available at-door.
+        It always offers a "Full Weekend Pass" if attendee badges are available.
+        If one-days are enabled, it will show either a generic "Single Day Pass" or a list of specific day badges,
+        depending on if PRESELL_ONE_DAYS is true. Each day's badge will need to be configured under [badge_ranges] and
+        [[badge]] to show up in this dropdown.
+        """
+        opts = []
+        if self.ATTENDEE_BADGE_AVAILABLE:
+            opts.append((self.ATTENDEE_BADGE, 'Full Weekend Pass (${})'.format(self.BADGE_PRICE)))
+        if self.ONE_DAYS_ENABLED and self.PRESELL_ONE_DAYS:
```
Comment (kitsuta) [inline uber/config.py]: What this does instead is it looks at the prices you have defined under the [[single_day]] portion of badge prices, then attempts to find badge types with matching names and add them to the dropdown.

So if you have:
```
[[single_day]]
Friday = 45
Saturday = 50
Sunday = 30
```

Then in order to make single-day badges work you need to add e.g. `friday_badge` to the [badge_ranges] section and the [[badge]] section under [enums]. As you can see, this code only executes if you have PRESELL_ONE_DAYS set to True - otherwise it goes back to the old behavior.

File: uber/config.py [hunk 2]
```diff
@@ -111,6 +111,15 @@ def get_attendee_price(self, dt):
     def get_group_price(self, dt):
         return self.get_attendee_price(dt) - self.GROUP_DISCOUNT
 
+    def get_badge_count_by_type(self, badge_type):
```
Comment (kitsuta) [inline uber/config.py]: In order to limit each day's badge type to a cap, I had to implement the ability to find things like `FRIDAY_BADGE_COUNT` on-the-fly, so we could automatically remove options as they sold out. This is what this function does, and it's called by the ever-expanding __getattr__ override, so all you need to do is set `friday_badge_stock` in your configuration to limit the number of Friday badges you sell.

File: uber/config.py [hunk 3]
```diff
@@ -165,8 +174,33 @@ def SUPPORTERS_ENABLED(self):
...
+                        badge = getattr(c, iterdate.strftime('%A').upper() + "_BADGE")
+                    except AttributeError:
+                        log.error('Could not add badge to AT_THE_DOOR_BADGE_OPTS. '
+                                  'Badge type not configured: {}', iterdate.strftime('%A').upper() + '_BADGE')
+                        iterdate += timedelta(days=1)
+                        continue
+
+                    if getattr(c, iterdate.strftime('%A').upper() + "_BADGE_AVAILABLE", None):
+                        opts.append((badge, iterdate.strftime('%A') + ' (${})'.format(price)))
+                iterdate += timedelta(days=1)
```
Comment (EliAndrewC) [inline uber/config.py]: Looks great!  Minor style suggestion: do ``try/except/else`` here to avoid needing to say ``continue`` and having to increment ``iterdate`` in 2 places.  In other words:

```python
    if price:
        try:
            badge = getattr(c, iterdate.strftime('%A').upper() + "_BADGE")
        except AttributeError:
            log.error('Could not add badge to AT_THE_DOOR_BADGE_OPTS. '
                      'Badge type not configured: {}', iterdate.strftime('%A').upper() + '_BADGE')
        else:
            if getattr(c, iterdate.strftime('%A').upper() + "_BADGE_AVAILABLE", None):
                opts.append((badge, iterdate.strftime('%A') + ' (${})'.format(price)))
    iterdate += timedelta(days=1)
```

File: uber/config.py (hunk 2 above)
Comment (EliAndrewC) [inline uber/config.py]: Semi-related side note since you mention our "everyexpanding ``__getattr__`` override: how do you feel about having to modify ``__getattr__`` every time to do this kind of stuff?  I'm cool with it, but it would also be fairly easy to implement a special type of property like ``derived_property`` to let us write code like

```python
@derived_property('BEFORE_')
def before(self, dt):
    return bool(dt and localized_now() > dt)

@derived_property('AFTER_')
def after(self, dt):
    return bool(dt and localized_now() < dt)

@derived_property('_COUNT')
def count(self, badge_type):
    with sa.Session() as session:
        return session.query(sa.Attendee).filter_by(badge_type=badge_type, status=c.COMPLETED_STATUS).count()
```

That kind of refactor is certainly out of scope for this PR, but I figured I'd check whether it's the general kind of thing you'd find appealing, or if you'd prefer to just keep adding elif clauses to ``__getattr__``.

File: uber/site_sections/registration.py [hunk 4]
```diff
@@ -500,7 +500,7 @@ def register(self, session, message='', **params):
                 message = 'Please select a payment type'
             elif attendee.payment_method == c.MANUAL and not re.match(c.EMAIL_RE, attendee.email):
                 message = 'Email address is required to pay with a credit card at our registration desk'
-            elif attendee.badge_type not in [c.ATTENDEE_BADGE, c.ONE_DAY_BADGE]:
+            elif attendee.badge_type not in [badge for badge, desc in c.AT_THE_DOOR_BADGE_OPTS]:
```
Comment (EliAndrewC) [inline uber/site_sections/registration.py]: This allows for the presale, but doesn't enforce the constraint of only being able to check in using badges on these days.  We also need to modify the ``Attendee.can_check_in`` property so that if someone shows up and tries to claim their ``Thursday`` badge on ``Saturday`` it won't let them without an admin changing the badge type and asking them for the difference in cost.  That should be like a one-line change, I'd expect.

**Comment by EliAndrewC** [comment]
Just wanted to say that this PR looks great.  I left a couple of minor suggestions, but overall your implementation is elegant and really well thought out - great work!

File: uber/config.py (hunk 2 above)
Comment (kitsuta) [inline uber/config.py]: I would LOVE a @derived_property property. The `__getattr__` methods are really cool and useful, but having them clumped together makes them hard to read.

File: uber/config.py (hunk 3 above)
Comment (kitsuta) [inline uber/config.py]: Oh jeez I totally forgot that `else` was an option here. Much more elegant.

File: uber/site_sections/registration.py (hunk 4 above)
Comment (kitsuta) [inline uber/site_sections/registration.py]: Hm, I think we may want to implement that in the magprime repo, particularly since these badge types don't actually have dates associated with them after you exit the loop in AT_THE_DOOR_BADGE_OPTS. Unless you think it's easy to re-associate them?

File: uber/config.py (hunk 2 above)
Comment (EliAndrewC) [inline uber/config.py]: Roger that, I've just created #1702 to track this for next year.

File: uber/site_sections/registration.py (hunk 4 above)
Comment (EliAndrewC) [inline uber/site_sections/registration.py]: Shouldn't be too hard to reassociate them.  We currently have ``c.BADGE_OPTS`` and ``c.BADGE_VARS`` properties which looks like this

```python
>>> c.BADGE_OPTS
[(51352218, 'Attendee'), (67489953, 'Supporter'), (16863825, 'Staff'), (144001217, 'Guest'), (71138939, 'One Day')]
```

So if we add these new badge types

```
[enums]
[[badge]]
thursday = "Thursday"
friday = "Friday"
saturday = "Saturday"
```

Then it'll look like

```python
>>> c.BADGE_OPTS
[(51352218, 'Attendee'), (67489953, 'Supporter'), (16863825, 'Staff'), (144001217, 'Guest'), (71138939, 'One Day'), (176673583, 'Thursday'), (84680830, 'Friday'), (62603428, 'Saturday')]
```

So we could rewrite ``can_check_in`` 

```python
@property
def can_check_in(self):
    return self.paid != c.NOT_PAID and self.badge_status in [c.NEW_STATUS, c.COMPLETED_STATUS] and not self.is_unassigned
```

to something like this

```python
@property
def can_check_in(self):
    valid = self.paid != c.NOT_PAID and self.badge_status in [c.NEW_STATUS, c.COMPLETED_STATUS] and not self.is_unassigned
    if valid and self.badge_type_label in {'Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'}:
        valid = self.badge_type_label == localized_now().strftime('%A')
    return valid
```

The only downside is that it requires that we MUST have the badge type labels match the day of week name.  I think that's fine, but it's something we'd want to document in ``configspec.ini``.

Comment (kitsuta) [inline uber/site_sections/registration.py]: Oh, okay. That sounds good.

**Comment by EliAndrewC** [comment]
Ok, this is ready to be merged.  I'll leave a bunch of comments inline explaining what I did, though I tried to add enough code comments that hopefully not much explanation is needed.

File: uber/config.py [hunk 5]
```diff
@@ -97,8 +97,10 @@ class Config(_Overridable):
     """
 
     def get_oneday_price(self, dt):
-        default = self.DEFAULT_SINGLE_DAY
-        return self.BADGE_PRICES['single_day'].get(dt.strftime('%A'), default)
+        return self.BADGE_PRICES['single_day'].get(dt.strftime('%A'), self.DEFAULT_SINGLE_DAY)
```
Comment (EliAndrewC) [inline uber/config.py]: The above is just a style refactor.

File: uber/config.py [hunk 6]
```diff
@@ -97,8 +97,10 @@ class Config(_Overridable):
...
+
+    def get_presold_oneday_price(self, badge_type):
+        return self.BADGE_PRICES['single_day'].get(self.BADGES[badge_type], self.DEFAULT_SINGLE_DAY)
```
Comment (EliAndrewC) [inline uber/config.py]: We need a separate function for "presold" single day passes, since their cost is based on badge type rather than the date on which they were registered.

File: uber/config.py [hunk 7]
```diff
@@ -133,10 +144,6 @@ def BADGE_PRICE(self):
         return self.get_attendee_price(sa.localized_now())
 
     @property
-    def SUPPORTER_BADGE_PRICE(self):
-        return self.BADGE_PRICE + self.SUPPORTER_LEVEL
-
-    @property
```
Comment (EliAndrewC) [inline uber/config.py]: This property hasn't been used anywhere in the codebase or any plugin for quite some time, so I removed it.

File: uber/config.py [hunk 8]
```diff
@@ -160,14 +167,29 @@ def PREREG_DONATION_TIERS(self):
         return dict(self.PREREG_DONATION_OPTS)
 
     @property
-    def SUPPORTERS_ENABLED(self):
-        return self.SUPPORTER_LEVEL in self.PREREG_DONATION_TIERS
-
-    @property
```
Comment (EliAndrewC) [inline uber/config.py]: Ditto for this property; it's not actually used in either the code plugin or our event-specific plugin.

File: uber/config.py [hunk 9]
```diff
@@ -175,10 +197,6 @@ def PREREG_AGE_GROUP_OPTS(self):
         return [opt for opt in self.AGE_GROUP_OPTS if opt[0] != self.AGE_UNKNOWN]
 
     @property
-    def DISPLAY_ONEDAY_BADGES(self):
-        return self.ONE_DAYS_ENABLED and sa.days_before(30, self.EPOCH)
```
Comment (EliAndrewC) [inline uber/config.py]: This option was actually never used; we put it in last year expecting that we'd turn on single day badge purchases a month before the event, but that was never implemented, so I removed the property.

File: uber/config.py [hunk 10]
```diff
@@ -302,6 +324,10 @@ def _unrepr(d):
 for _opt, _val in chain(_config.items(), c.BADGE_PRICES.items()):
     if not isinstance(_val, dict) and not hasattr(c, _opt.upper()):
         setattr(c, _opt.upper(), _val)
+for _opt, _val in c.BADGE_PRICES['stocks'].items():
+    _opt = _opt.upper() + '_STOCK'
+    if not hasattr(c, _opt):
+        setattr(c, _opt, _val)
```
Comment (EliAndrewC) [inline uber/config.py]: See ``configspec.ini`` below for an explanation, but I added a new ``[[stocks]]`` config section where we can track how many of each badge type we have available.  As per the ``c.*_AVAILABLE`` implementation, it's totally optional, but it allows us to automatically cut off selling badges of each type after a certain number.

File: uber/config.py [hunk 11]
```diff
@@ -319,11 +345,10 @@ def _is_intstr(s):
...
 
-'''
-Under certain conditions, we want to completely remove certain payment options from the system.
-However, doing so cleanly also risks an exception being raised if these options are referenced elsewhere in the code
-(i.e., c.STRIPE). So we create an enum val to allow code to check for these variables without exceptions.
-'''
+
+# Under certain conditions, we want to completely remove certain payment options from the system.
+# However, doing so cleanly also risks an exception being raised if these options are referenced elsewhere in the code
+# (i.e., c.STRIPE). So we create an enum val to allow code to check for these variables without exceptions.
```
Comment (EliAndrewC) [inline uber/config.py]: Just a style fix.

File: uber/configspec.ini [hunk 12]
```diff
@@ -131,11 +131,6 @@ collect_exact_birthdate = boolean(default=False)
 # emergency contact number, and cellphone number.
 collect_full_address = boolean(default=False)
 
-# An event may or may not want attendees to volunteer through pre-reg - this controls
-# the visibility of the volunteering checkbox.
-# TODO: Replace this with a larger per-form visibility configuration for all special fields
-volunteer_form_visible = boolean(default=False)
```
Comment (EliAndrewC) [inline uber/configspec.ini]: This hasn't been used in quite some time if it ever was, so I removed it.

File: uber/templates/registration/register.html [hunk 13]
```diff
@@ -25,9 +23,14 @@
...
-        removeVolunteerRows();
-        $('#bold-field-message').detach().prependTo('.form-horizontal');
+        {% if c.AT_THE_DOOR_BADGE_OPTS %}
+            maybeBold();
+            removeDonationRows();
+            removeVolunteerRows();
+            $('#bold-field-message').detach().prependTo('.form-horizontal');
+        {% else %}
+            $('.panel').empty().append('<h1>All badges are sold out!</h1>');
+        {% endif %}
```
Comment (EliAndrewC) [inline uber/templates/registration/register.html]: If I thought this had any serious chance of actually happening, I'd put more effort into styling it.  Or maybe I wouldn't and would instead spend my time rolling in money in the treasury room and making it rain.

In case it's not clear, the moment we run out of all badge types sold at the door, this page stops presenting and form and instead just displays a big header message saying we're all sold out.

File: uber/config.py [hunk 14]
```diff
@@ -394,6 +419,18 @@ def _is_intstr(s):
...
+if c.ONE_DAYS_ENABLED and c.PRESELL_ONE_DAYS:
+    _day = c.EPOCH.date()
+    while _day <= c.ESCHATON.date():
+        _name = _day.strftime('%A')
+        _val = c.create_enum_val(_name)
+        c.BADGES[_val] = _name
+        c.BADGE_OPTS.append((_val, _name))
+        c.BADGE_VARS.append(_name.upper())
+        c.BADGE_RANGES[_val] = c.BADGE_RANGES[c.ONE_DAY_BADGE]
+        _day += timedelta(days=1)
```
Comment (EliAndrewC) [inline uber/config.py]: What I'm about to explain is documented in the comments in ``configspec.ini`` but I'll repeat it here because this is a slight change from what I proposed in an earlier comment in the discussion on this ticket.

If you set ``presell_one_days`` to true, this code automatically inserts a badge type for each day of the event.  In out case that means we have ``c.THURSDAY``, ``c.FRIDAY``, ``c.SATURDAY``, and ``c.SUNDAY`` badge types automatically configured (based on the ``c.EPOCH`` and ``c.ESCHATON`` settings).  We do NOT have to add new options to the ``[[badge]]`` config subsection of the ``[enums]`` config section.

Comment (EliAndrewC) [inline uber/config.py]: Also in case it's not clear, all of these single day passes share the same badge range as the ``c.ONE_DAY_BADGE`` type.

File: uber/templates/registration/register.html [hunk 15]
```diff
@@ -16,8 +16,15 @@
...
-
+    var maybeWarn = function () {
+        var types = {{ c.PRESOLD_ONEDAY_BADGE_TYPES|jsonize }};
+        var badgeType = $.val('badge_type');
+        if (badgeType in types && types[badgeType] !== moment().format('dddd')) {
+            $('#day-warning').text('This badge can ONLY be picked up on ' + types[badgeType] + '.  You cannot pick it up today or any other day than ' + types[badgeType] + '.');
+        } else {
+            $('#day-warning').empty();
+        }
+    };
```
Comment (EliAndrewC) [inline uber/templates/registration/register.html]: We originally talked about putting this check in the ``magprime`` repo, but this seemed generic enough that it could go into the core repo.

File: uber/config.py (hunk 8 above)
Comment (kitsuta) [inline uber/config.py]: Pretty sure this was for magstock.

File: uber/config.py (hunk 14 above)
Comment (kitsuta) [inline uber/config.py]: ... did you really need to make a config option to list the days of the week? :P

Comment (kitsuta) [inline uber/config.py]: More seriously, although this will work for us, I don't LOVE that it's less flexible than what I had programmed originally. Then again, I can't really say "make people configure things right" is a great alternative either.

File: uber/configspec.ini (hunk 12 above)
Comment (kitsuta) [inline uber/configspec.ini]: This was used for AnthroCon, but that was pre-event-specific-plugins.

**Comment by kitsuta** [comment]
+1 - thanks for finishing this up for me!

File: uber/config.py (hunk 8 above)
Comment (EliAndrewC) [inline uber/config.py]: That may be why we originally added it, but I think it's no longer used even for that, because we're currently doing different if-else checks which don't involve this property.

File: uber/config.py (hunk 14 above)
Comment (EliAndrewC) [inline uber/config.py]: Yeah, the reason why I ended up doing it this way is that the original approach basically required an event which wanted to presell single day passes to add a set of configuration options VERY PRECISELY.  As in, if they added an option like ``friday = "Friday Pass"`` then everything would have broken, because the option MUST be ``friday = "Friday"`` due to how we're doing the lookups.

That kind of restriction is just begging for weird, hard-to-diagnose bugs, which is why I went this route.
